### PR TITLE
reduct-cpp: add recipe

### PR DESCRIPTION
**Pull request by toge**
Specify library name and version:  **reduct-cpp/***

Apple-clang doesn't support reduct-cpp due to `operator<=>` support.
https://developer.apple.com/xcode/cpp/#c++20

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/reduct-cpp/all/patches/1.9.0-0001-use-cci.patch [hunk 1]
```diff
@@ -0,0 +1,41 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index cda9d1d..e6cf1d8 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -14,9 +14,20 @@ set(REDUCT_CPP_ENABLE_TESTS OFF CACHE BOOL "Compile tests")
+ set(CMAKE_CXX_STANDARD 20)
```
Comment (RubenRBS) [inline recipes/reduct-cpp/all/patches/1.9.0-0001-use-cci.patch:6]: This should be commented out for the `compiler.cppstd` setting to take effect :)

**Review by RubenRBS — COMMENTED** [review]
Hi! Thanks a lot, only a minor change, otherwise this looks super good :)

File: recipes/reduct-cpp/all/patches/1.9.0-0001-use-cci.patch (hunk 1 above)
Comment (toge) [reply recipes/reduct-cpp/all/patches/1.9.0-0001-use-cci.patch]: @RubenRBS 
Thank you for your review!
Fixed.

File: recipes/reduct-cpp/all/test_package/test_package.cpp [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#include "reduct/client.h"
+
+#include <iostream>
+
+using reduct::IBucket;
+using reduct::IClient;
+
+int main() {
+    auto client = IClient::Build("http://127.0.0.1:8383");
```
Comment (RubenRBS) [inline recipes/reduct-cpp/all/test_package/test_package.cpp:9]: Can we avoid hitting the network layer, even if it is just locally? Thanks!

Comment (toge) [reply recipes/reduct-cpp/all/test_package/test_package.cpp:11]: @RubenRBS 
Thank you for your comment.
Fixed.

File: recipes/reduct-cpp/all/patches/1.9.0-0001-use-cci.patch (hunk 1 above)
Comment (toge) [reply recipes/reduct-cpp/all/patches/1.9.0-0001-use-cci.patch]: @RubenRBS 
I have a question.
Setting to use C++20 is defined by upstream.
Should we remove such definition from upstream codes on all recipes?

**Comment by atimin** [comment]
Hi @toge , the creator of the library here. I really appropriate your effort to integrate `reduct-cpp` into the `conan` infrastructure. If you need any support from my side, feel free to ask.

**Comment by toge** [comment]
@atimin 
Thank you for follow this PR!
I'm trying to pass all CI environments include Windows.
After it has been solved, I will create several PR  to upstream which solvesr portability issues.

**Comment by AbrilRBS** [comment]
Let me know if you need help getting msvc support sorted, but this might be a nice case for validating out such configuration if the errors persis :)

File: recipes/reduct-cpp/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    def layout(self):
+        cmake_layout(self, src_folder="src")
+
+    def requirements(self):
+        self.requires("openssl/[>=1.1 <4]")
+        self.requires("fmt/11.0.2")
+        self.requires("cpp-httplib/0.18.0")
+        self.requires("nlohmann_json/3.11.3")
+        self.requires("concurrentqueue/1.0.4")
+        self.requires("date/3.0.1")
```
Comment (uilianries) [inline recipes/reduct-cpp/all/conanfile.py:62]:
```suggestion
        if not self.options.with_std_chrono:
            self.requires("date/3.0.1")
```

If I understand correctly, date is only required when not using std::chrono 

https://github.com/reductstore/reduct-cpp/blob/main/cmake/InstallDependencies.cmake#L56

File: recipes/reduct-cpp/all/patches/1.12.0-0001-use-cci.patch [hunk 4]
```diff
@@ -0,0 +1,39 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index af71f65..7e3a0e2 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -16,7 +16,20 @@ set(CMAKE_CXX_STANDARD 20)
+ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+ set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
+ 
+-include(cmake/InstallDependencies.cmake)
++find_package(fmt REQUIRED CONFIG)
```
Comment (uilianries) [inline recipes/reduct-cpp/all/patches/1.12.0-0001-use-cci.patch:10]: What situation! We need to patch it in order to not use Conan! 

https://github.com/reductstore/reduct-cpp/blob/main/cmake/InstallDependencies.cmake#L12

File: recipes/reduct-cpp/all/patches/1.12.0-0004-include-compare.patch
Comment (uilianries) [inline recipes/reduct-cpp/all/patches/1.12.0-0004-include-compare.patch:1]: This patch could be sent to the upstream.

**Review by uilianries — COMMENTED** [review]
Thank you for your PR! 😄

**Comment by AbrilRBS** [comment]
Seems like this has issues compiling for Windows

**Comment by victor1234** [comment]
@toge I want to finalize this PR. Could you let me know what else needs to be fixed and what the difficulties are?

**Comment by toge** [comment]
@victor1234 
Thank you for your interest in this PR!
Due to conan-center-index introducing the limit of 30 PRs per person, I had to closed this PR.

If you are interested in making a package of reduct-cpp, you may copy the codes of this PR and create other PR.

Or would you prefer that I reopen this PR?
Fortunately I have one PR slot left now, I can resume this PR.

**Comment by toge** [comment]
@victor1234 
How about the situation?
I am not particular about my recipes and you may copy the content of my PR.
Or would you prefer that I develop my PR again?

**Comment by victor1234** [comment]
@toge I missed the notifications about your replies. Both options work for me. If you’d like to finish it, feel free to reopen the PR and I’ll complete it. If you prefer to keep the slot, I can copy the code and open a new PR.

**Pull request closed without merging**